src/odemis/driver/tfsbc.py

```diff
@@ -0,0 +1,376 @@
+        """
+        if len(values) != 4:
+            raise ValueError("write_registers received payload of invalid length %s != 4." % len(values))
+
+        # Check if values are in allowed range
+        if not all(0 <= val <= 0xFFFF for val in values):
+            raise ValueError("Register values %s not in range [0, 65535]." % values)
+
+        try:
+            # write all registers together (starting at lower x register (=0x01))
+            rq = self._serial.write_registers(BEAMDEFL_LX, values, unit=SLAVE_UNIT)
+        except IOError:
+            self._reconnect()
+            raise IOError("Failed to write registers of beam control firmware, "
+                          "restarted serial connection.")
+
+    def _read_registers(self):
+        """
+        Read all four registers. Try to reconnect to device in case connection was lost.
+        :return (list of 4 ints): register values (-x, -y, x, y)
+        """
+        try:
+            # write all registers together (starting at lower x register (=0x01))
+            rr = self._serial.read_holding_registers(BEAMDEFL_LX, 4, unit=SLAVE_UNIT)
+            return rr.registers
+        except IOError:
+            self._reconnect()
+            raise IOError("Failed to write registers of beam control firmware, "
+                          "restarted serial connection.")
+
+    def _reconnect(self):
+        """
+        Attempt to reconnect. It will block until this happens.
+        On return, the hardware should be ready to use as before.
+        """
+        num_it = 5
+        self.state._set_value(model.HwError("Beam deflection controller disconnected"), force_write=True)
+        logging.warning("Failed to write registers, trying to reconnect...")
+        for i in range(num_it):
+            try:
+                self._serial.close()
+                self._serial = None
+                self._port = self._findDevice(self._portpattern, self._serialnum)
+                self._serial = self._openSerialPort(self._port)
+                logging.info("Recovered device.")
+                break
+            except IOError:
+                continue
+        else:
+            raise IOError("Failed to reconnect to beam deflection controller.")
+        self.state._set_value(model.ST_RUNNING, force_write=True)
+
+    def updateMetadata(self, md):
+        if model.MD_CALIB_BEAMSHIFT in md:
+            # Check format
+            bs = md[model.MD_CALIB_BEAMSHIFT]
+            try:
+                if not len(bs) == 4:  # 4 tuples required
+                    raise ValueError("Invalid MD_CALIB_BEAMSHIFT metadata %s: 4 tuples required." % (bs,))
+                if not all(len(val) == 2 for val in bs):  # each of the 4 values is a tuple of 2
+                    raise ValueError("Invalid MD_CALIB_BEAMSHIFT metadata %s: Two values per tuple required." % (bs,))
+                if not all(all([type(val) in (int, float) for val in tup]) for tup in bs):  # each element is a number
```

Comment:
You don't need to put a `[]` around the generator.
You shouldn't use `type` to check the class, but `isinstance()`, which also handles the case of sub-classes (eg, numpy.float).
So you could write something like:
`if not all(all(isinstance(val, (int, float)) for val in tup) for tup in bs):`

Note that you could even write, but it's not readable:
`if not all(isinstance(val, (int, float)) for tup in bs for val in tup):`